Kernel/Storage/ATA/AHCIController.cpp

```diff
@@ -26,6 +26,21 @@ UNMAP_AFTER_INIT NonnullRefPtr<AHCIController> AHCIController::initialize(PCI::D
 
 bool AHCIController::reset()
 {
+    SpinlockLocker locker(m_control);
+    Array<u32, 32> lock_flags {};
+
+    for (auto port : m_ports) {
+        if (port)
+            lock_flags[port->m_port_index] = port->m_hard_lock.lock();
+    }
+
+    ScopeGuard release_all_spinlocks([this, lock_flags]() {
+        for (auto port : m_ports) {
+            if (port)
+                port->m_hard_lock.unlock(lock_flags[port->m_port_index]);
+        }
+    });
```

Comment:
This looks really strange. Is it really necessary to lock all the individual ports across the entire `reset()` operation like this? Would it not suffice to lock the controller here, and lock each port as we process it?

Reply:
The idea is to prevent individual AHCI ports to try to initiate a transaction as this can be _fatal_ (this is UB if I recall correctly). I keep the transaction initiation logic in each AHCIPort and not putting it in the AHCIController as it will impact performance of other attached devices severely. If this is needed, I'd like to add a comment explaining this.

Reply:
I added a comment and hopefully this is enough to explain why we do this.

Reply:
I think the misunderstanding here is that you confuse this method (which does a full AHCI reset, i.e. HBA reset) - not actually resetting each AHCIPort (we could do that and in fact we do that when we initializing an AHCIPort).